File: mobile-app/lib/models/nguoi_dung.dart
/// Model thong tin nguoi dung dang nhap (tra ve tu GET /auth/me)
class NguoiDung {
  final String id;
  final String tenDangNhap;
  final String? hoTen;
  final bool quyenNhapLieu;
  final bool quyenUploadTaiLieu;
  final String phamViXem; // "gioi_han" | "toan_bo"
  final bool quyenXuatBaoCao;
  final bool isAdmin;
  final String? thonPhuTrachId;

  NguoiDung({
    required this.id,
    required this.tenDangNhap,
    this.hoTen,
    required this.quyenNhapLieu,
    required this.quyenUploadTaiLieu,
    required this.phamViXem,
    required this.quyenXuatBaoCao,
    required this.isAdmin,
    this.thonPhuTrachId,
  });

  factory NguoiDung.fromJson(Map<String, dynamic> json) {
    return NguoiDung(
      id: json['id']?.toString() ?? '',
      tenDangNhap: json['ten_dang_nhap']?.toString() ?? '',
      hoTen: json['ho_ten']?.toString(),
      quyenNhapLieu: json['quyen_nhap_lieu'] == true,
      quyenUploadTaiLieu: json['quyen_upload_tai_lieu'] == true,
      phamViXem: json['pham_vi_xem']?.toString() ?? 'gioi_han',
      quyenXuatBaoCao: json['quyen_xuat_bao_cao'] == true,
      isAdmin: json['is_admin'] == true,
      thonPhuTrachId: json['thon_phu_trach_id']?.toString(),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'ten_dang_nhap': tenDangNhap,
        'ho_ten': hoTen,
        'quyen_nhap_lieu': quyenNhapLieu,
        'quyen_upload_tai_lieu': quyenUploadTaiLieu,
        'pham_vi_xem': phamViXem,
        'quyen_xuat_bao_cao': quyenXuatBaoCao,
        'is_admin': isAdmin,
        'thon_phu_trach_id': thonPhuTrachId,
      };
}

File: backend/app/models/nguoi_dung.py
import uuid
from datetime import datetime

from sqlalchemy import Boolean, Enum, ForeignKey, String
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column
from sqlalchemy.sql import func

from app.db.base import Base

pham_vi_xem_enum = Enum("gioi_han", "toan_bo", name="pham_vi_xem_enum", create_type=False)


class NguoiDung(Base):
    __tablename__ = "nguoi_dung"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    ten_dang_nhap: Mapped[str] = mapped_column(String(50), nullable=False, unique=True)
    mat_khau_hash: Mapped[str] = mapped_column(String(255), nullable=False)
    ho_ten: Mapped[str] = mapped_column(String(150), nullable=False)
    chuc_vu: Mapped[str | None] = mapped_column(String(150), nullable=True)
    so_dt: Mapped[str | None] = mapped_column(String(20), nullable=True)
    email: Mapped[str | None] = mapped_column(String(150), nullable=True)
    anh_dai_dien: Mapped[str | None] = mapped_column(String(255), nullable=True)

    is_admin: Mapped[bool] = mapped_column(Boolean, default=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)

    quyen_nhap_lieu: Mapped[bool] = mapped_column(Boolean, default=False)
    quyen_upload_tai_lieu: Mapped[bool] = mapped_column(Boolean, default=False)
    pham_vi_xem: Mapped[str] = mapped_column(pham_vi_xem_enum, default="gioi_han")
    quyen_xuat_bao_cao: Mapped[bool] = mapped_column(Boolean, default=False)
    # chi_lap_bien_ban: cot da co trong migration 0004 nhung CHUA duoc ap dung vao DB
    # (thieu quyen postgres superuser). Tam thoi comment de tranh loi UndefinedColumnError
    # lam hong toan bo dang nhap/API nguoi_dung. Bo comment sau khi chay xong migration 0004.
    # chi_lap_bien_ban: Mapped[bool] = mapped_column(Boolean, default=False)

    thon_phu_trach_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("thon.id"), nullable=True
    )

    created_at: Mapped[datetime] = mapped_column(server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(server_default=func.now())
